src/cargo/core/compiler/build_context/target_info.rs

```diff
@@ -748,3 +751,79 @@ impl RustcTargetData {
         self.target_config(kind).links_overrides.get(lib_name)
     }
 }
+
+/// Structure used to deal with Rustdoc fingerprinting
+#[derive(Debug, Serialize, Deserialize)]
+pub struct RustDocFingerprint {
+    pub rustc_vv: String,
+}
+
+impl RustDocFingerprint {
+    /// Read the `RustDocFingerprint` info from the fingerprint file.
+    fn read<'a, 'cfg>(cx: &Context<'a, 'cfg>) -> CargoResult<Self> {
+        let rustdoc_data = paths::read(&cx.files().host_root().join(".rustdoc_fingerprint.json"))?;
+        serde_json::from_str(&rustdoc_data).map_err(|e| anyhow::anyhow!("{:?}", e))
+    }
+
+    /// Write the `RustDocFingerprint` info into the fingerprint file.
+    fn write<'a, 'cfg>(&self, cx: &Context<'a, 'cfg>) -> CargoResult<()> {
+        paths::write(
+            &cx.files().host_root().join(".rustdoc_fingerprint.json"),
+            serde_json::to_string(&self)?.as_bytes(),
+        )
+    }
+
+    fn remove_doc_dirs(doc_dirs: &Vec<&Path>) -> CargoResult<()> {
+        doc_dirs
+            .iter()
+            .filter(|path| path.exists())
+            .map(|path| paths::remove_dir_all(&path))
+            .filter(|res| res.is_err())
+            .collect::<CargoResult<Vec<()>>>()
+            .map(|_| ())
```

Comment:
Sorry, just one more minor change and I think this should be good to go.  You can collect directly to a `Result` since it implements `FromIterator`, stopping at the first `Err`.

```suggestion
            .collect::<CargoResult<()>>()
```